### PR TITLE
Separate interpreter and engine per wazero.cache impl

### PR DESCRIPTION
**Pull request by mathetake**
Signed-off-by: Takeshi Yoneda <takeshi@tetrate.io>

### CODE REVIEW

File: .github/workflows/commit.yaml [hunk 1]
```diff
@@ -79,7 +79,7 @@ jobs:
         if: ${{ github.event_name  == 'pull_request' }}
 
       # Run tests with -race only on main branch push.
-      - run: make test go_test_options='-race'
+      - run: make test go_test_options='-timeout 10m -race'
```
Comment (mathetake) [inline .github/workflows/commit.yaml:82]: turns out that 2m was not enough for main branch race tests

File: cache_test.go [hunk 2]
```diff
@@ -22,11 +24,16 @@ func TestCompilationCache(t *testing.T) {
 		foo, bar := getCacheSharedRuntimes(ctx, t)
 		cacheInst := foo.cache
 
+		eng := foo.cache.engs[engineKindInterpreter]
```
Comment (codefromthecrypt) [inline cache_test.go:27]:
```suggestion
		// add interpreter first, to ensure compiler support isn't order dependent
		eng := foo.cache.engs[engineKindInterpreter]
```

File: config.go [hunk 3]
```diff
@@ -145,6 +145,14 @@ var engineLessConfig = &runtimeConfig{
 	dwarfDisabled:         false,
 }
 
+type engineKind int
+
+const (
+	engineKindCompiler engineKind = iota
+	engineKindInterpreter
+	engineKindNum
```
Comment (codefromthecrypt) [inline config.go:153]: supernit: engineKindCount?

**Review by codefromthecrypt — APPROVED** [review]
thanks for doing this. sometimes we tell people to temporarily use interpreter so I got nervous before.

**Comment by mathetake** [comment]
🤞  on the timeout stuff on main branch test

**Pull request merged**